Reinstall feature

AWolf81 opened the pull request:
**Related Issue:**
#278

**Summary:**
- Added application menu item Reinstall dependencies to Current Project.
- Added Reinstall to dependency Saga:
   - Removes Node_modules folder
   - Triggers `yarn` call
   - Watches stdout from yarn for status updates to LoadingScreen

**Things to discuss:**
- Do we need a confirmation before executing the reinstall? I think it's not needed as nothing will be destroyed and it's pretty clear what the click will do
- Speech bubble (hidden by default, in the screencast below it only shows because I had it displayed in the previous reinstall): Should we remove it or leave it as is? I just wanted to show a progress but it maybe useful to keep for advanced users.
- For the watching of the stdout we could extract the createStdioChannel function from `task.saga` but I don't  like the API from it. I think we could refactor this afterwards. What do you think?
- Progressbar: It needs to be improved because current fix of 100% display is hacky. Is there a way to get more outputs about the progress from Yarn?

**Todos**
- [x] Resolve conflict for `app-status.reducer`
- [x] Fix failing tests
- [x] Check if progress bar can be improved. Improved the progress bar so it can not jump back (like in the screenrecording below).:
  - 100% fix done with a delay (a bit hacky but works)
  - Animating back on next reinstall run fixed with `react-spring` (ProgressBar component picked from migration branch) & added reset property
  - Jumping backwards fixed by summing the progress instead of using absolute values.

**Screenshots/GIFs:**
**Screenrecording of reinstall**
![screenrecording_reinstall_27112018](https://user-images.githubusercontent.com/3046542/49063848-06fabd00-f219-11e8-8cab-7d3beee7262d.gif)

**Application menu item**
![grafik](https://user-images.githubusercontent.com/3046542/47868882-b359a700-de05-11e8-8425-1c5facd472cf.png)

**Install Dependencies button (if node_modules missing)**
![grafik](https://user-images.githubusercontent.com/3046542/48400263-2b896c00-e726-11e8-8308-939a02406256.png)

melanieseltzer commented on src/components/LoadingScreen/LoadingScreen.js:
```diff
@@ -1,22 +1,76 @@
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.statusText !== this.props.statusText) {
+      // new status
+      const [_, total] = (/\d+\/\d+/.exec(nextProps.statusText) || [
```
nit: _ is unused

AWolf81 replied:
`_` is a placeholder because I need the second value. 
Maybe I'll change this to use the index of the regex result.

melanieseltzer replied:
How about this? https://stackoverflow.com/a/44560070

AWolf81 replied:
Yes, I've tried this too. It's working but I don't like how it looks.
I think I've seen the `_` destructuring in the Guppy code base before but it's better to avoid it. 
If changed it so it should be OK now.

melanieseltzer commented on src/components/LoadingScreen/LoadingScreen.js:
```diff
@@ -1,22 +1,76 @@
 // @flow
-import React, { PureComponent } from 'react';
+import React, { Fragment, PureComponent } from 'react';
```
nit: Fragment is unused

AWolf81 replied:
Oh, OK, I missed that one. I'll remove it.

melanieseltzer commented on src/components/LoadingScreen/LoadingScreen.js:
```diff
@@ -1,22 +1,76 @@
-    const { showLoadingScreen } = this.props;
+    const { showLoadingScreen, statusText } = this.props;
+    const { showStatus } = this.state;
+    const ellipsis = statusText.length > 220 ? '...' : '';
```
I'm getting undefined for `statusText` so the loading screen never moves beyond the first message in the speech bubble. However it looks like the reinstall is working because I deleted my node_modules to test, and they successfully re-appear. 

<img width="400" alt="screen shot 2018-11-02 at 7 36 23 am" src="https://user-images.githubusercontent.com/17421347/47921567-2567d580-de72-11e8-8082-d49e85a92a2e.png">

<img width="400" alt="screen shot 2018-11-02 at 7 28 00 am" src="https://user-images.githubusercontent.com/17421347/47921056-e7b67d00-de70-11e8-99c0-080e04950f49.png">

It successfully exits:

<img width="460" alt="screen shot 2018-11-02 at 7 38 49 am" src="https://user-images.githubusercontent.com/17421347/47921659-61029f80-de72-11e8-9dac-6285887a4220.png">

AWolf81 replied:
OK, I don't get that here. 
But I'll add a conditional or pass a default `statusText` prop so it won't fail.

AWolf81 replied:
Please try again. It should work now.

melanieseltzer commented:
So, I've been trying to investigate a really weird performance issue in this branch and I'm confused because I can't pinpoint what's causing it. I've checked out each commit one by one to see if it gets introduced somewhere, and it appears and disappears at random.

The top menu will not allow me to click on anything, I roll over something and it just closes the menu. Also there's some weird jumpiness going on with scrolling (page jumps up and down).

<img src="https://user-images.githubusercontent.com/17421347/48041454-1d3dbd80-e132-11e8-8c4c-778db743e171.gif" width="600">

This also happens with the reinstall (no errors to console). It flickers and loading screen disappears, but reinstall does eventually work and dependencies reappear. But the flicker is strange.

<img src="https://user-images.githubusercontent.com/17421347/48041777-a30e3880-e133-11e8-8ff2-97ad3824f82b.gif" width="600">

I'm stumped. I can't tell if it's my machine or widespread, do you encounter any of this on your machine? :/ As far as I can tell it's not on master and just in this branch.

AWolf81 commented:
I don't have that issue. Have you tried to remove `node_modules` folder from Guppy and reinstall everything.
Really a weird issue. I'll check it on Ubuntu maybe there is a similar behaviour.

joshwcomeau commented:
Ah-ha - when I open Guppy and open the Redux Devtools, I see "LOAD_DEPENDENCY_INFO_FROM_DISK" firing over and over. Looks like there's some recursive loop somewhere.

joshwcomeau commented on src/components/ProgressBar/ProgressBar.js:
```diff
@@ -11,21 +11,30 @@ type Props = {
 
 class ProgressBar extends Component<Props> {
   static defaultProps = {
+    position: 'inline',
```
I don't think `position: inline` is a thing. Did you mean `display: inline`?

AWolf81 replied:
Maybe the naming here is a bit off. It's not a css style. It's intended as a switch prop between having the progress bar inline or on the top edge of the screen.
I'd like to have it on the top edge for the loading screen.

joshwcomeau commented on src/components/ProgressBar/ProgressBar.js:
```diff
@@ -45,9 +54,10 @@ class ProgressBar extends Component<Props> {
 
 const Wrapper = styled.div`
-  position: relative;
+  position: ${props => (props.position === 'top' ? 'absolute' : 'relative')};
```
I think I'd rather that the ProgressBar just fill its parent container. It feels weird to me that the progress bar controls its own positioning. I'd do something like this instead:

```js
// in LoadingScreen.js
<ProgressBarWrapper>
  <ProgressBar />
</ProgressBarWrapper>

const ProgressBarWrapper = styled.div`
  position: absolute;
  top: 0;
left: 0;
right: 0;
`
```

AWolf81 replied:
OK, that's better. I've modified it & removed the `position` prop.

joshwcomeau commented on src/reducers/app-status.reducer.js:
```diff
@@ -2,32 +2,55 @@
+      return {
+        ...state,
+        statusText: newStatus !== '' ? newStatus : state.statusText, // only update if string defined
+      };
```
We probably want to avoid creating a new object if we don't have anything to change. So we could do:

```js
if (newStatus === '') {
  return state;
}

return {
  ...state,
  statusText: newStatus,
}
```

The rationale is that redux reducers should return the existing object when nothing changes, to `PureComponent`s won't re-render when they don't need to.

AWolf81 replied:
Oh, good point. I'll change that.

joshwcomeau commented on src/sagas/dependency.saga.js:
```diff
@@ -108,6 +153,28 @@ export function* handleUninstallDependenciesStart({
+        // Yield exit code and complete stdout
+        yield output;
+
+        console.log('exit', output);
```
Nit: console.log

joshwcomeau commented on src/components/ProjectPage/ProjectPage.js:
```diff
@@ -55,7 +56,11 @@ class ProjectPage extends PureComponent<Props> {
+    const { project } = this.props;
+    if (
+      project.id !== nextProps.project.id ||
+      project.dependencies.length === 0
```
Ah I think this is the problem.

I'm guessing that when there are 0 dependencies, it triggers the `loadDependencyInfoFromDisk`, which re-renders its parent, which re-invokes this method.

I think we should handle the zero-dependency case separately. Maybe we can use the AppStatus reducer to check if it's in the middle of doing an install before invoking the method?

melanieseltzer commented on src/components/ProjectPage/ProjectPage.js:
```diff
@@ -59,10 +61,10 @@ class ProjectPage extends PureComponent<Props> {
     if (
       project.id !== nextProps.project.id ||
-      project.dependencies.length === 0
+      (project.dependencies.length === 0 && !isReinstallingActive)
```
I think this is supposed to be `isReinstallingActive` not `!isReinstallingActive` right? As in, if there's no dependencies and we triggered the reinstall, then reinstall. Otherwise the way it is now, it's saying if there's no dependencies and reinstall isn't active (aka always!)... reinstall infinitely. 

Definitely the cause of the perf issues - thanks @joshwcomeau for finding it! A gif of the nastiness in action 😮

<img width="300" alt="screen shot 2018-11-11 at 8 09 31 am" src="https://user-images.githubusercontent.com/17421347/48315423-e11fb800-e58a-11e8-9e8d-d169b230bc1b.png">

<img width="400" src="https://user-images.githubusercontent.com/17421347/48315424-e67d0280-e58a-11e8-866a-c4c3662519e0.gif">

AWolf81 replied:
My idea was only load if no reinstalling is currently active. I'll try the console.log later.

melanieseltzer replied:
I think I get what you're trying to do with that but problem with that condition is if you're in the app not clicking the reinstall button, then you're technically in that state always (you're always having no dependencies and you're always not in reinstall state), so `loadDependencyInfoFromDisk` is being triggered infinitely.

AWolf81 replied:
Ah I see, I should also get that problem if I'm removing `node_modules` folder manually in explorer. I think we should add a flow so that the loading happens only once. 
Maybe we have to add a check that will stop the loading if the folder doesn't exists.

I need to check why it's looping.

AWolf81 commented:
@joshwcomeau Thanks for your review.
The performance issue should be fixed. Please check if it's fixed.
I rewrote the `dependency.reducer` and added the loading status per project there.
Also I'm hiding all tasks if `node_modules` folder is missing as I think that's easier than disabling every button in the UI. What do you think? Is it OK? I think it's OK as this really an edge case and shouldn't happen often.

**There are two points open:**
- We probably need a better copy for the info text if `node_modules` folder is missing. See updated screenshot.
- I'll check if I can add some animations to the speech bubble - if not I think it's also OK to remove it as I think the progress bar is enough info. Do you have an idea how I could fix the display of the progress bar so 100% is visible?

AWolf81 commented on src/components/ProjectPage/ProjectPage.js:
```diff
@@ -136,9 +136,9 @@ class ProjectPage extends PureComponent<Props> {
-                Please install the dependencies so you can use the scripts that
-                are needed by your project. This is OK and happens if you're
-                freshly cloning a project from Github.
+                Please install dependencies so you can use the scripts that are
```
Thanks for the new copy. That's better 👍 I'll update it

AWolf81 replied:
OK, I added your copy. I also changed the styling a bit.
I'm also using enum-style conditional rendering instead of `&&` conditionals in `ProjectPage`. I think we have used rendering methods before but I like that's less code. 
What do you think is it OK to use?

melanieseltzer replied:
Great new look, I love it with the icon 😄 

Enum rendering is interesting, I don't think I've come across it before. LGTM, since it's just a code style thing I think it's fine to use, can always be refactored later if necessary. But yeah I do agree it seems cleaner.

AWolf81 commented on src/components/ProjectPage/ProjectPage.js:
```diff
@@ -58,7 +62,11 @@ class ProjectPage extends PureComponent<Props> {
+    const { project } = this.props;
+    if (
+      project.id !== nextProps.project.id ||
+      project.dependencies.length === 0
```
Not exactly sure if this caused the looping. But I'll try if it can be removed. I think a update is required if there are now dependencies in state so reload from disk should be triggered. So I think it can't be removed but maybe we can only reload if `loadedStatus` in `dependencies.reducer` is `loading`. Maybe we have to add a state like `in progress` so this can be prevented from retriggering loading.

For me the looping stopped by removing the swallowing of the EONET error. But if you're still having it I'll check that line.

AWolf81 replied:
OK, I've removed the `project.dependencies.length` check and couldn't detect a difference. Or am I missing a case? 
Please check if the issue is fixed with this.

melanieseltzer replied:
Works great now :-) Yay! 

RE: updating if there are new dependencies in state. From how I understand the new code, triggering a reinstall via clicking the button should trigger the reload from disk/UI update from this code in dependency.saga.js:

```js
// load dependencies to refresh state
yield put(loadDependencyInfoFromDiskStart(projectId, projectPath));
```

But @joshwcomeau reports things still not working for him properly so maybe I'm wrong.

AWolf81 commented on src/reducers/dependencies.reducer.js:
```diff
@@ -1,7 +1,10 @@
+  LOAD_DEPENDENCY_INFO_FROM_DISK_ERROR,
   LOAD_DEPENDENCY_INFO_FROM_DISK_FINISH,
   ADD_DEPENDENCY,
   UPDATE_DEPENDENCY,
```
Yeah, I know. I've kept it because I don't like to use `any` for the action type.
But I don't know why it's not working with the type. Do you have an idea? I don't remember the type error right now. But the only fix was to use `any`.

AWolf81 commented on src/sagas/dependency.saga.js:
```diff
@@ -100,6 +111,49 @@ export function* handleInstallDependenciesStart({
+    const channel = yield call(reinstallDependencies, projectPath);
+
+    // The channel is used to pass every termianl output to loadingScreen status text
+    // todo: Check if we need to pass a progress value so we can display a progress bar as well.
```
Yes, you're right. I'll remove it.

AWolf81 commented:
@joshwcomeau can you please have a look at this? Is the performance OK now?

I've added everything from your review and removed the speech bubble. There is a new issue for the bubble so we can add this later. I think we can mark that issue as low priority as I think it's not that important to add.

j-f1 commented on src/components/ProjectPage/ProjectPage.js:
```diff
@@ -110,18 +125,58 @@ class ProjectPage extends PureComponent<Props> {
+                  </Paragraph>
+                  <Paragraph>
+                    This can happen if you've freshly cloned a project from
+                    Github. In order to run scripts in Guppy, you'll need to
```
```suggestion
                    GitHub. In order to run scripts in Guppy, you'll need to
```

melanieseltzer commented:
@joshwcomeau Is this still blocking? Perf issues seem resolved and UI messages are hidden for now until UI is worked on, I think we can merge...
